### PR TITLE
add initial implementation of defender start/stop functions

### PR DESCRIPTION
**Pull request by qiutongs**
Description
-----------

- add initial implementation for defender start/stop functions 
    - use two timers (this might be a temporary approach. TBD)
- MQTT connect/disconnect

Checklist:
----------

- [X] Build succeed
- [NO] I have tested my changes. No regression in existing tests.
- [X] My code is Linted.

### CODE REVIEW

File: lib/defender/aws_iot_defender.c [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        /* metrics publish timer must be created and armed */
+        /* guard timer must be created */
+        if( AwsIotClock_TimerCreate( &_metricsPublishTimer, _metricsPublishTimerExpirationRoutine, NULL ) )
+        {
+            if( AwsIotClock_TimerArm( &_metricsPublishTimer, _periodMilliSecond, _periodMilliSecond ) )
+            {
+                if( AwsIotClock_TimerCreate( &_guardTimer, _guardTimerExpirationRoutine, NULL ) )
+                {
+                    defenderError = AWS_IOT_DEFENDER_SUCCESS;
+                }
```
Comment (lt72) [inline lib/defender/aws_iot_defender.c]: missing else?

File: lib/defender/aws_iot_defender.c [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+/*-----------------------------------------------------------*/
+
+AwsIotDefenderError_t AwsIotDefender_Start( AwsIotDefenderStartInfo_t * pStartInfo )
+{
+    AwsIotDefenderError_t defenderError = AWS_IOT_DEFENDER_INTERNAL_FAILURE;
+
+    /* if current state is not started */
+    if( !isStarted() )
```
Comment (lt72) [inline lib/defender/aws_iot_defender.c]: not thread safe... that is by design, correct?

File: lib/defender/aws_iot_defender.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        }
+    }
+    else
+    {
+        defenderError = AWS_IOT_DEFENDER_ALREADY_STARTED;
+    }
+
+    if( defenderError != AWS_IOT_DEFENDER_SUCCESS )
+    {
+        _startInfo = _START_INFO_RESETER;
```
Comment (lt72) [inline lib/defender/aws_iot_defender.c]: two Ts, I think...

File: lib/defender/aws_iot_defender.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+{
+    AwsIotDefenderError_t defenderError = AWS_IOT_DEFENDER_INTERNAL_FAILURE;
+
+    if( periodSeconds < AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS )
+    {
+        defenderError = AWS_IOT_DEFENDER_PERIOD_TOO_SHORT;
+    }
+    else
+    {
+        _periodMilliSecond = periodSeconds * 1000;
```
Comment (lt72) [inline lib/defender/aws_iot_defender.c]: Check for overflow, although unlikely...

File: lib/defender/aws_iot_defender.c (hunk 1 above)
Comment (qiutongs) [reply lib/defender/aws_iot_defender.c]: No. The return value is initialized AWS_IOT_DEFENDER_INTERNAL_FAILURE. So by default, this function return "internal failure"

File: lib/defender/aws_iot_defender.c (hunk 2 above)
Comment (qiutongs) [reply lib/defender/aws_iot_defender.c]: Yeah, I plan to deal with thread safety for all in the end.

File: lib/defender/aws_iot_defender.c (hunk 3 above)
Comment (qiutongs) [reply lib/defender/aws_iot_defender.c]: Yeah, Let me fix in next PR

File: lib/defender/aws_iot_defender.c (hunk 4 above)
Comment (qiutongs) [reply lib/defender/aws_iot_defender.c]: Yes, I will add an upper limit

**Pull request merged**